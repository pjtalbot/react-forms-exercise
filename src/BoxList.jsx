import React, { useState } from "react"
import Box from "./Box.jsx"
import NewBoxForm from "./NewBoxForm.jsx"

// BoxList - Place your state that contains all of the boxes here. This component should render all of the Box components along with the NewBoxForm component

const BoxList = () => {
    const [boxes, setBoxes] = useState([]);
    const addBox = boxObj => {
        setBoxes(boxes => [...boxes, boxObj]);
    }
    const remove = id => {
        setBoxes(boxes => boxes.filter(box => box.id !== id))
    };

    const allBoxComponents = boxes.map(box => {
        return (
            <Box 
        key={box.id}
        id={box.id}
        width={box.width}
        height={box.height}
        handleRemove={remove}
        backgroundColor={box.backgroundColor}
        />

        )
        
    })

    return (
        <div className="BoxList">
        <NewBoxForm createBox={addBox}/>
        {allBoxComponents}
        </div>
    )
        
}


export default BoxList;