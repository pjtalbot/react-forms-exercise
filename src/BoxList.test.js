import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";
import BoxList from "./BoxList";

function addBox(boxList, height = "2", width = "2", backgroundColor = "peachpuff") {
    const heightInput = boxList.getByLabelText("Box Height");
    const widthInput = boxList.getByLabelText("Box Width");
    const backgroundInput = boxList.getByLabelText("Box Color");
    fireEvent.change(backgroundInput, { target: { value: backgroundColor} });
    fireEvent.change(widthInput, { target: { value: width } });
    fireEvent.change(heightInput, { target: { value: height } });
    const button = boxList.getByText("Add!");
    fireEvent.click(button);
  }

  it("renders the list", function() {
    render(<BoxList />)
  })

  it("matches snapshot", function() {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
  });
  

  it("can add a new box", function() {
    const boxList = render(<BoxList />);
  
    // no boxes yet
    expect(boxList.queryByText("X")).not.toBeInTheDocument();
  
    addBox(boxList);
  
    // expect to see a box
    const removeButton = boxList.getByText("X");
    expect(removeButton).toBeInTheDocument();
  
    // expect(asFragment()).toMatchSnapshot();
  });

  it("can remove a box", function() {
    const boxList = render(<BoxList />);
    addBox(boxList);
  
    const removeButton = boxList.getByText("X");
  
    // click the remove button and the box should be gone
    fireEvent.click(removeButton);
    expect(removeButton).not.toBeInTheDocument();
  });