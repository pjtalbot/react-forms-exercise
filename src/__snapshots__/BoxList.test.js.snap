// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <div>
      <form>
        <div>
          <label
            for="width"
          >
            Box Width
          </label>
          <input
            id="width"
            name="width"
            type="text"
            value="200"
          />
        </div>
        <div>
          <label
            for="height"
          >
            Box Height
          </label>
          <input
            id="height"
            name="height"
            type="text"
            value="200"
          />
        </div>
        <div>
          <label
            for="backgroundColor"
          >
            Box Color
          </label>
          <input
            id="backgroundColor"
            name="backgroundColor"
            type="text"
            value="black"
          />
        </div>
        <button
          id="newBoxButton"
        >
          Add!
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
